From 592755295cb1d0292ff423f1640a548181c1b4f6 Mon Sep 17 00:00:00 2001
From: Gregory Norton <gregory@serenade.local>
Date: Sat, 30 Aug 2026 00:00:00 +0000
Subject: [PATCH 6/7] LibKeyboard: Tolerate a missing kernel keymap on host systems

setkeymap()/getkeymap() are Serenity LibC wrappers around kernel syscalls.
On host systems there is no kernel keymap; report ENOSYS instead of
failing to build (serenity.h does not exist in the host libc).

diff --git a/Userland/Libraries/LibKeyboard/CharacterMap.cpp b/Userland/Libraries/LibKeyboard/CharacterMap.cpp
index 9b6cf63215..539d98aa55 100644
--- a/Userland/Libraries/LibKeyboard/CharacterMap.cpp
+++ b/Userland/Libraries/LibKeyboard/CharacterMap.cpp
@@ -5,10 +5,13 @@
  */
 
 #include "CharacterMap.h"
+#include <AK/Platform.h>
 #include <AK/StringBuilder.h>
 #include <LibKeyboard/CharacterMapFile.h>
 #include <errno.h>
-#include <serenity.h>
+#ifdef AK_OS_SERENITY
+#    include <serenity.h>
+#endif
 
 namespace Keyboard {
 
@@ -27,11 +30,17 @@ CharacterMap::CharacterMap(ByteString const& map_name, CharacterMapData const& m
 
 int CharacterMap::set_system_map()
 {
+#ifdef AK_OS_SERENITY
     return setkeymap(m_character_map_name.characters(), m_character_map_data.map, m_character_map_data.shift_map, m_character_map_data.alt_map, m_character_map_data.altgr_map, m_character_map_data.shift_altgr_map);
+#else
+    // There is no kernel keymap to update on host systems.
+    return -ENOSYS;
+#endif
 }
 
 ErrorOr<CharacterMap> CharacterMap::fetch_system_map()
 {
+#ifdef AK_OS_SERENITY
     CharacterMapData map_data;
     char keymap_name[50 + 1] = { 0 };
 
@@ -39,6 +48,10 @@ ErrorOr<CharacterMap> CharacterMap::fetch_system_map()
         return Error::from_errno(errno);
 
     return CharacterMap { keymap_name, map_data };
+#else
+    // There is no kernel keymap to fetch on host systems.
+    return Error::from_errno(ENOSYS);
+#endif
 }
 
 ByteString const& CharacterMap::character_map_name() const
